Dynamic Content Scaling resource cache cannot cache images between layers

hortont424 opened the pull request:
#### e8c0db1780a3059ff32170d89dfe8be43aed1937
<pre>
Dynamic Content Scaling resource cache cannot cache images between layers
<a href="https://bugs.webkit.org/show_bug.cgi?id=245920">https://bugs.webkit.org/show_bug.cgi?id=245920</a>
<a href="https://rdar.apple.com/100336143">rdar://100336143</a>

Reviewed by Said Abou-Hallawa.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::allocateImageBuffer):
(WebKit::RemoteRenderingBackend::ensureDynamicContentScalingResourceCache):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingBifurcatedImageBuffer.h:
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingBifurcatedImageBuffer.mm:
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingImageBufferBackend.mm:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::allocateBuffer const):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.mm:
(WebKit::RemoteLayerTreeContext::ensureDynamicContentScalingResourceCache):
* Source/WebCore/platform/graphics/re/DynamicContentScalingResourceCache.h: Added.
(WebCore::DynamicContentScalingResourceCache::DynamicContentScalingResourceCache):
* Source/WebCore/platform/graphics/re/DynamicContentScalingResourceCache.mm: Added.
(WebCore::DynamicContentScalingResourceCache::create):
Move CGDisplayList&apos;s image cache object up to RemoteLayerTreeContext (in the in-process case)
and RemoteRenderingBackend (in the GPU-process case), so that resources can be cached
across multiple layers.

* Source/WebCore/platform/graphics/ImageBuffer.h:
Note that it is not safe in a unified-sources world to forward-declare WebCore::IOSurface
because of the name conflict with ::IOSurface; we need to strictly order the declarations,
which is easiest to achieve by just importing our header.

Canonical link: <a href="https://commits.webkit.org/273240@main">https://commits.webkit.org/273240@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/434a27564d5e9796c36bb354cd79644a59931877

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34828 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13653 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36841 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37516 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31429 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35972 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16062 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10738 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30378 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35357 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/11574 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31014 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10107 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/10186 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/31091 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38781 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31632 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31418 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/36222 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10287 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8196 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34191 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12115 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7989 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10832 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11179 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingBifurcatedImageBuffer.h:
```diff
@@ -50,6 +50,8 @@ class DynamicContentScalingBifurcatedImageBuffer : public WebCore::ImageBuffer {
 
     WebCore::GraphicsContext& context() const final;
 
+    static WebCore::DynamicContentScalingResourceCache createResourceCache();
```
Why is this function part of this class? Can't `DynamicContentScalingResourceCache` have its own header and implementation files?

hortont424 replied:
I suppose it could!

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.h:
```diff
@@ -54,6 +54,10 @@ class IOSurfacePool;
 
+#if ENABLE(RE_DYNAMIC_CONTENT_SCALING)
+using DynamicContentScalingResourceCache = RetainPtr<id>;
+#endif
```
Why is this definition here? Should not it be defined in its own header file?

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.h:
```diff
@@ -65,10 +69,13 @@ struct ImageBufferCreationContext {
     IOSurfacePool* surfacePool { nullptr };
     PlatformDisplayID displayID { 0 };
+#endif
+#if ENABLE(RE_DYNAMIC_CONTENT_SCALING)
+    DynamicContentScalingResourceCache dynamicContentScalingResourceCache;
```
Not new. But I am really confused by the individual parameters and `ImageBufferCreationContext`. I do not know when adding a new parameter, should it be passed directly to `ImageBuffer::create()`? Or should it be a member of `ImageBufferCreationContext`?

Maybe (in a future patch) we can rename `ImageBufferCreationContext` `ImageBufferPlatformCreationContext` or `ImageBufferBackendCreationContext`. Just to tell us what parameters should be in this struct.

hortont424 replied:
Makes sense!

shallawa commented on Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingImageBufferBackend.mm:
```diff
@@ -84,7 +84,7 @@ static CFDictionaryRef makeContextOptions(const DynamicContentScalingImageBuffer
-    , m_resourceCache(bridge_id_cast(adoptCF(RECGCommandsCacheCreate(nullptr))))
+    , m_resourceCache(creationContext.dynamicContentScalingResourceCache)
     , m_renderingMode(renderingMode)
 {
 }
```
Shouldn't we assert here `ASSERT(m_resourceCache);`? Or is it okay to have `DynamicContentScalingImageBufferBackend` with a null `m_resourceCache`?

hortont424 replied:
It's fine to have one without a resource cache, it'll just send ports for images every time.

shallawa commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
```diff
@@ -276,8 +276,12 @@ RefPtr<ImageBuffer> RemoteRenderingBackend::allocateImageBuffer(const FloatSize&
+    if (m_gpuConnectionToWebProcess->isDynamicContentScalingEnabled() && (purpose == RenderingPurpose::LayerBacking || purpose == RenderingPurpose::DOM)) {
+        if (!m_dynamicContentScalingResourceCache)
+            m_dynamicContentScalingResourceCache = DynamicContentScalingBifurcatedImageBuffer::createResourceCache();
+        creationContext.dynamicContentScalingResourceCache = m_dynamicContentScalingResourceCache;
```
Cannot this be done through an ensure function?
```
creationContext.dynamicContentScalingResourceCache = ensureDynamicContentScalingResourceCache();
```

shallawa replied:
Similar to what is done in `RemoteLayerTreeContext::dynamicContentScalingResourceCache()`.

hortont424 replied:
Sure, don't see why not!

shallawa commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
```diff
@@ -236,8 +236,11 @@
-    if (m_parameters.includeDisplayList == IncludeDisplayList::Yes)
+    if (m_parameters.includeDisplayList == IncludeDisplayList::Yes) {
+        if (auto* context = m_layer->context())
+            creationContext.dynamicContentScalingResourceCache = context->dynamicContentScalingResourceCache();
```
Is it okay to create `DynamicContentScalingBifurcatedImageBuffer` with a null `dynamicContentScalingResourceCache`? Should not we have a fallback and create a non shared resource case in this case?
```
if (auto* context = m_layer->context())
    creationContext.dynamicContentScalingResourceCache = context->dynamicContentScalingResourceCache();
else
    creationContext.dynamicContentScalingResourceCache = DynamicContentScalingBifurcatedImageBuffer::createResourceCache();
```

hortont424 replied:
No, this is fine. There's minimal point in making a non-shared non-persistent cache.

hortont424 committed:
Dynamic Content Scaling resource cache cannot cache images between layers
https://bugs.webkit.org/show_bug.cgi?id=245920
rdar://100336143

Reviewed by Said Abou-Hallawa.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::allocateImageBuffer):
(WebKit::RemoteRenderingBackend::ensureDynamicContentScalingResourceCache):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingBifurcatedImageBuffer.h:
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingBifurcatedImageBuffer.mm:
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingImageBufferBackend.mm:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::allocateBuffer const):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.mm:
(WebKit::RemoteLayerTreeContext::ensureDynamicContentScalingResourceCache):
* Source/WebCore/platform/graphics/re/DynamicContentScalingResourceCache.h: Added.
(WebCore::DynamicContentScalingResourceCache::DynamicContentScalingResourceCache):
* Source/WebCore/platform/graphics/re/DynamicContentScalingResourceCache.mm: Added.
(WebCore::DynamicContentScalingResourceCache::create):
Move CGDisplayList's image cache object up to RemoteLayerTreeContext (in the in-process case)
and RemoteRenderingBackend (in the GPU-process case), so that resources can be cached
across multiple layers.

* Source/WebCore/platform/graphics/ImageBuffer.h:
Note that it is not safe in a unified-sources world to forward-declare WebCore::IOSurface
because of the name conflict with ::IOSurface; we need to strictly order the declarations,
which is easiest to achieve by just importing our header.

Canonical link: https://commits.webkit.org/273240@main
